### PR TITLE
Add Wavefront support

### PR DESCRIPTION
**Pull request by jkschneider**
See https://github.com/micrometer-metrics/micrometer/issues/365.

Adds support for VMWare's dimensional monitoring system, Wavefront.

cc / @howardyoo

### CODE REVIEW

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/export/signalfx/SignalFxMetricsExportAutoConfigurationTests.java [hunk 1]
```diff
@@ -50,7 +49,7 @@
 	@Test
 	public void backsOffWithoutAClock() {
 		this.runner.run((context) -> assertThat(context)
-				.doesNotHaveBean(InfluxMeterRegistry.class));
+				.doesNotHaveBean(SignalFxMeterRegistry.class));
```
Comment (jkschneider) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/export/signalfx/SignalFxMetricsExportAutoConfigurationTests.java]: Nit on similar SignalFx test that was referring to the wrong type.

Comment (wilkinsona) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/export/signalfx/SignalFxMetricsExportAutoConfigurationTests.java]: Oops. Thanks for the catch.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+public class WavefrontProperties extends StepRegistryProperties {
+	/**
+	 * The URI to publish metrics to. The URI could represent a Wavefront sidecar or the
+	 * Wavefront API host. This host could also represent an internal proxy set up in your environment
+	 * that forwards metrics data to the Wavefront API host.
+	 *
+	 * If publishing metrics to a Wavefront proxy (as described in https://docs.wavefront.com/proxies_installing.html),
+	 * the host must be in the proxy://HOST:PORT format.
+	 */
+	private String uri;
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java]: Could this be a `URI`? The same question applies to Atlas, Influx, New Relic and SignalFX but I missed those previously.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontMetricsExportAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for exporting metrics to StatsD.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontMetricsExportAutoConfiguration.java]: `StatsD` -> `Wavefront`

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	/**
+	 * Uniquely identifies the app instance that is publishing metrics to Wavefront. Defaults to the local host name.
+	 */
+	private String source;
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java]: Whatever we decide to do with SignalFX's source, we should do the same thing here.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	 * Required when publishing directly to the Wavefront API host, otherwise does nothing.
+	 */
+	private String apiToken;
+
+	/**
+	 * Wavefront metrics are grouped hierarchically by name in the UI. Setting a global prefix separates
+	 * metrics originating from this app's whitebox instrumentation from those originating from other Wavefront
+	 * integrations.
+	 */
+	private String globalPrefix;
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java]: The first sentence in the javadoc is what's used in the metadata. To improve what's shown in an IDE, I think I'd reword this to something like the following:

> Global prefix to separate metrics originating from this app's white box instrumentation from those originating from other Wavefront integrations when viewed in the Wavefront UI.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java (hunk 2 above)
Comment (jkschneider) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java]: I tested this, and yes, `URI` would work fine. I could change the type on the underlying interfaces (`InfluxConfig`, `NewRelicConfig`, etc.) for all _but_ `AtlasConfig`, which is a type that we don't control. So I think for consistency I'll leave them as String in Micrometer, but there should be no problem mapping a URI in Spring props to a String on the corresponding Micrometer interface.

I could change the PR to URI here or we can merge this as a String and I can follow up with a single PR that changes all of them to URI. Whichever you prefer.

Comment (wilkinsona) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java]: Let's go with URI for this one, and then make all the others use `URI` in a separate commit/PR.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java (hunk 4 above)
Comment (jkschneider) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/wavefront/WavefrontProperties.java]: We decided to leave it as is I think.
